### PR TITLE
Marketplace CLI commands

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Adds the following commands that utilize the Marketplace feature:
- `marketplace:search` - Accepts a term and returns any available plugin with that term in its key, name or description(s). If no term is specified, all available plugins are listed. The result is shown in a table with the key, name, and short description (Best match for their current language).
- `marketplace:info` - Accepts a plugin's key and returns the full information about the plugin provided by the Marketplace API.
- `marketplace:download` - Accepts one or more comma-separated plugin keys and downloads the latest compatible versions. This never installs the plugin (use the dedicated command for that) and by default, it will only download if it hasn't been downloaded already (Use --force for updates).

These commands were added with the `marketplace` prefix instead of `plugin` just so it is clear that they use the Marketplace feature.

### CODE REVIEW

File: src/Console/Marketplace/DownloadCommand.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+        $this->addArgument('plugins', InputArgument::REQUIRED, 'The internal plugin name(s)');
+        $this->addOption('force', 'f', InputOption::VALUE_NONE, 'Force download even if the plugin is already downloaded');
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $plugins = $input->getArgument('plugins');
+        if (!empty($plugins)) {
+            $plugins = explode(' ', $plugins);
```
Comment (trasher) [inline src/Console/Marketplace/DownloadCommand.php:61]: You talk about comma-separated values in PR description but rely on spaces
```suggestion
            $plugins = explode(',', $plugins);
```

Comment (cedric-anne) [reply src/Console/Marketplace/DownloadCommand.php:61]: Or maybe rely on `InputArgument::IS_ARRAY` flag, to ensure you get an array of values.

**Review by trasher — APPROVED** [review]

File: src/Console/Marketplace/DownloadCommand.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+class DownloadCommand extends AbstractCommand
+{
+    protected function configure()
+    {
+        parent::configure();
+
+        $this->setName('glpi:marketplace:download');
+        $this->setAliases(['marketplace:download']);
+        $this->setDescription('Download plugin from the GLPI marketplace');
```
Comment (cedric-anne) [inline src/Console/Marketplace/DownloadCommand.php:51]:
```suggestion
        $this->setDescription(__('Download plugin from the GLPI marketplace'));
```

File: src/Console/Marketplace/DownloadCommand.php [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        $this->addArgument('plugins', InputArgument::REQUIRED | InputArgument::IS_ARRAY, 'The internal plugin name(s)');
```
Comment (cedric-anne) [inline src/Console/Marketplace/DownloadCommand.php:53]:
```suggestion
        $this->addArgument('plugins', InputArgument::REQUIRED | InputArgument::IS_ARRAY, __('The plugin key'));
```

File: src/Console/Marketplace/DownloadCommand.php [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        $this->addOption('force', 'f', InputOption::VALUE_NONE, 'Force download even if the plugin is already downloaded');
```
Comment (cedric-anne) [inline src/Console/Marketplace/DownloadCommand.php:54]:
```suggestion
        $this->addOption('force', 'f', InputOption::VALUE_NONE, __('Force download even if the plugin is already downloaded'));
```

File: src/Console/Marketplace/DownloadCommand.php [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        $plugins = $input->getArgument('plugins');
+
+        foreach ($plugins as $plugin) {
+            if (!empty(trim($plugin))) {
```
Comment (cedric-anne) [inline src/Console/Marketplace/DownloadCommand.php:62]: Code seems less complex and easier to read if you do not encapsulate the whole `foreach` code inside a `if` block.
```php
if (empty(trim($plugin))) {
    continue;
}
```

File: src/Console/Marketplace/DownloadCommand.php [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+                // If the plugin is already downloaded, refuse to download it again
+                if (!$input->getOption('force') && is_dir(GLPI_MARKETPLACE_DIR . '/' . $plugin)) {
+                    $error_msg = sprintf('Plugin "%s" is already downloaded. Use --force to force it to re-download.', $plugin);
```
Comment (cedric-anne) [inline src/Console/Marketplace/DownloadCommand.php:65]:
```suggestion
                    $error_msg = sprintf(__('Plugin "%s" is already downloaded. Use --force to force it to re-download.'), $plugin);
```

File: src/Marketplace/Controller.php [hunk 7]
```diff
@@ -86,7 +86,7 @@ public static function getAPI(): PluginsApi
      *
      * @return bool
      */
-    public function downloadPlugin(): bool
+    public function downloadPlugin($skip_install = false): bool
```
Comment (cedric-anne) [inline src/Marketplace/Controller.php:89]: Double negatives are often subject to issues.
I propose to use `$auto_install = true`.

Also, you should force the parameter type to boolean and add corresponding doc.

File: src/Console/Marketplace/DownloadCommand.php [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                    $result = $controller->downloadPlugin(true);
+                    $success_msg = sprintf(__("Plugin %s downloaded successfully"), $plugin);
+                    $error_msg = sprintf(__("Plugin %s could not be downloaded"), $plugin);
+                    if ($result) {
+                        $output->writeln("<info>$success_msg</info>");
+                    } else {
+                        $output->writeln("<error>$error_msg</error>");
+                    }
+                } else {
+                    throw new \RuntimeException(sprintf(__('Plugin "%s" cannot be downloaded'), $plugin));
```
Comment (cedric-anne) [inline src/Console/Marketplace/DownloadCommand.php:80]: For a more user friendly output.
```suggestion
                    $output->writeln('<error>' . sprintf(__('Plugin "%s" cannot be downloaded'), $plugin) . '</error>');
                    return 1;
```

File: src/Console/Marketplace/InfoCommand.php [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+class InfoCommand extends AbstractCommand
+{
+    protected function configure()
+    {
+        parent::configure();
+
+        $this->setName('glpi:marketplace:info');
+        $this->setAliases(['marketplace:info']);
+        $this->setDescription('Get information about a plugin');
```
Comment (cedric-anne) [inline src/Console/Marketplace/InfoCommand.php:53]:
```suggestion
        $this->setDescription(__('Get information about a plugin'));
```

File: src/Console/Marketplace/InfoCommand.php [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+        $this->addArgument('plugin', InputArgument::REQUIRED, 'The plugin key');
```
Comment (cedric-anne) [inline src/Console/Marketplace/InfoCommand.php:55]:
```suggestion
        $this->addArgument('plugin', InputArgument::REQUIRED, __('The plugin key'));
```

File: src/Console/Marketplace/InfoCommand.php [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+        $controller = new Controller();
+        $plugins = $controller::getAPI()->getAllPlugins();
+
+        $result = array_filter($plugins, static function ($p) use ($plugin) {
+            return strtolower($p['key']) === strtolower($plugin);
+        });
+
+        if (count($result) === 0) {
+            $output->writeln('<error>Plugin not found</error>');
```
Comment (cedric-anne) [inline src/Console/Marketplace/InfoCommand.php:72]:
```suggestion
            $output->writeln('<error>' . __('Plugin not found') . '</error>');
```

File: src/Console/Marketplace/SearchCommand.php [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+
+class SearchCommand extends AbstractCommand
+{
+    protected function configure()
+    {
+        parent::configure();
+
+        $this->setName('glpi:marketplace:search');
+        $this->setAliases(['marketplace:search']);
+        $this->setDescription('Search GLPI marketplace');
```
Comment (cedric-anne) [inline src/Console/Marketplace/SearchCommand.php:53]:
```suggestion
        $this->setDescription(__('Search GLPI marketplace'));
```

File: src/Console/Marketplace/SearchCommand.php [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+        $this->addArgument('term', InputArgument::OPTIONAL, 'The search term');
```
Comment (cedric-anne) [inline src/Console/Marketplace/SearchCommand.php:55]:
```suggestion
        $this->addArgument('term', InputArgument::OPTIONAL, __('The search term'));
```

**Review by cedric-anne — APPROVED** [review]

**Pull request merged**